packages/vite/src/node/optimizer/optimizer.ts

```diff
@@ -160,7 +182,7 @@ async function createDepsOptimizer(
   // On warm start or after the first optimization is run, we use a simpler
   // debounce strategy each time a new dep is discovered.
   let crawlEndFinder: CrawlEndFinder | undefined
-  if (isBuild || !cachedMetadata) {
+  if ((!noDiscovery && isBuild) || !cachedMetadata) {
```

Comment:
When `noDiscovery: true`, we don't need the `crawlEndFinder`. This simplifies the logic in `onCrawlEnd`.